### PR TITLE
Improve status for unpushed branches

### PR DESCRIPTION
**Pull request by greg0ire**
This is a cleaned up, rebased, resurrection of #2085

### CODE REVIEW

File: src/Composer/Downloader/GitDownloader.php [hunk 1]
```diff
@@ -120,6 +143,13 @@ protected function cleanChanges(PackageInterface $package, $path, $update)
     {
         GitUtil::cleanEnv();
         $path = $this->normalizePath($path);
+
+        if ($this instanceof DvcsDownloaderInterface) {
```
Comment (stof) [inline src/Composer/Downloader/GitDownloader.php]: this makes no sense. It will always be true

Comment (greg0ire) [inline src/Composer/Downloader/GitDownloader.php]: @stof: indeed!
@asgrim : any rationale behind this ? Can I safely remove it ?

Comment (asgrim) [inline src/Composer/Downloader/GitDownloader.php]: No idea, I guess so! :)

Comment (greg0ire) [inline src/Composer/Downloader/GitDownloader.php]: @stof : fixed

**Comment by naderman** [comment]
What's the upside of this over https://github.com/composer/composer/issues/2253#issuecomment-39662835 ?

**Comment by greg0ire** [comment]
From the description of #2553 this does not only work for branches that have diverged, but for changes branches that are behind their remote counterpart (but maybe the description is bad). Plus, it gives a better feedback with what exactly is wrong with status codes : there are different status codes for different situations (changes commited but not pushed, changes not commited, (changes commited and not pushed in one vendor && changes not commited in another)).

**Comment by greg0ire** [comment]
Also, it is more generic : there is an interface that is introduced so that the same functionality can be implemented for mercurial, bazaar and your favorite DCVS.

**Comment by greg0ire** [comment]
I just rebased, the tests are green.

**Comment by greg0ire** [comment]
ping @naderman are there any plans to merge this (or #2553) ?

**Comment by Seldaek** [comment]
Thanks @greg0ire, sorry for the delay. I had to do quite a few tweaks though (see https://github.com/composer/composer/commit/be4d385942cce9b0d1116aba985a99380c7a53d1) to avoid false positives for me.. Hopefully it'll go well in the wild and I didn't miss anything.

**Comment by greg0ire** [comment]
Better late than never I guess :) Thanks for merging that!

**Pull request merged**